### PR TITLE
spark history server for paytmlabs

### PR DESCRIPTION
**Pull request by viperKiran**
1. We are going to support only AWS and S3
2. Refactored entrypoint.sh script to support step1
3. Refactored Dockerfile to support step 1

Automated docker build CI is enabled - https://hub.docker.com/r/paytmcanada/spark-history-server-docker/tags

verified the image with history-server deployment and it works

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -1,187 +1,21 @@
...
-  fi;
-else
-    export SPARK_HISTORY_OPTS="$SPARK_HISTORY_OPTS \
-    -Dspark.history.fs.logDirectory=$eventsDir";
-fi;
+export SPARK_HISTORY_OPTS="$SPARK_HISTORY_OPTS \
+  -Dspark.history.fs.logDirectory=${eventsDir} \
+  -Dspark.hadoop.fs.s3a.impl=org.apache.hadoop.fs.s3a.S3AFileSystem \
+  -Dspark.hadoop.fs.s3a.server-side-encryption-algorithm=${serverSideEncryption} \
+  -Dspark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.EnvironmentVariableCredentialsProvider";
```
Comment (maxbrunet) [inline entrypoint.sh]: Why not use the default credential provider chain?
```suggestion
  -Dspark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.DefaultAWSCredentialsProviderChain";
```
BTW isn't it the default?

File: Dockerfile [hunk 2]
```diff
@@ -1,20 +1,14 @@
-ARG SPARK_IMAGE=gcr.io/spark-operator/spark:v2.4.0
+ARG SPARK_IMAGE=gcr.io/spark-operator/spark:v3.0.0-hadoop3
 FROM ${SPARK_IMAGE}
 
-# Set up dependencies for Google Cloud Storage access.
-RUN rm $SPARK_HOME/jars/guava-14.0.1.jar
-ADD http://repo1.maven.org/maven2/com/google/guava/guava/23.0/guava-23.0.jar $SPARK_HOME/jars
-# Add the connector jar needed to access Google Cloud Storage using the Hadoop FileSystem API.
-ADD https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-latest-hadoop2.jar $SPARK_HOME/jars
+USER root
```
Comment (maxbrunet) [inline Dockerfile]: Why this statement?

The base image seems to already have it:
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/v1beta2-1.2.0-3.0.0/Dockerfile#L38

Or is this one? If yes, could you restore it at the end of the Dockerfile?
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/v1beta2-1.2.0-3.0.0/Dockerfile.rh#L56

Apart from the build, is the image able to run as a non-root user?

File: entrypoint.sh [hunk 3]
```diff
@@ -1,187 +1,21 @@
...
-# If there is no passwd entry for the container UID, attempt to create one
-if [[ -z "${uid_entry}" ]] ; then
-    if [[ -w /etc/passwd ]] ; then
-        echo "$uid:x:$uid:$gid:anonymous uid:${SPARK_HOME}:/bin/false" >> /etc/passwd
-    else
-        echo "Container entrypoint.sh failed to add passwd entry for anonymous UID"
-    fi
-fi
+readonly eventsDir="${1}"
+readonly serverSideEncryption="${2}"
```
Comment (maxbrunet) [inline entrypoint.sh]: Please use uppercase and snake case for constants.
I would also stay close to the name of the spark property:

```suggestion
readonly SPARK_LOG_DIR="${1}"
readonly S3_SSE_ALGORITHM="${2}"
```

**Review by maxbrunet — CHANGES_REQUESTED** [review]
I wonder if this `entrypoint.sh` is really necessary. The most important are the AWS libraries. Wouldn't it be simpler to just set the spark command in the Kubernetes deployment? This would be less maintenance for us, and more flexibility for everyone.

File: entrypoint.sh (hunk 1 above)
Comment (viperKiran) [reply entrypoint.sh]: Nope, I don't want to use the default. The default is simple with access and secret keys. But we can go with a secret object and environment in this way.

Comment (maxbrunet) [reply entrypoint.sh]: The default provider chain chains all the others and `EnvironmentVariableCredentialsProvider` is the first in the chain:
https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default

File: Dockerfile (hunk 2 above)
Comment (viperKiran) [reply Dockerfile]: Never tried with non-root user but looked at the jars its all deployed with root users. Not sure what root 185 stands for here. 
On other hand, I can't copy jars (hadoop and aws-sdk) into SPARK_HOME/jars as a non-root user.

File: entrypoint.sh (hunk 3 above)
Comment (viperKiran) [reply entrypoint.sh]: Sounds good, I will go with your suggestion.

**Comment by viperKiran** [comment]
moving this PR to new repo https://github.com/PaytmLabs/docker-spark-history-server/pull/1

**Pull request closed without merging**